### PR TITLE
consul service discovery for envoy gateway

### LINKED ISSUE #1561: Consul service discovery for envoy gateway
### How will you implement it

---

### PR DESCRIPTION
**Pull request by rumeshmadhusanka**
### Purpose

Consul service discovery for envoy gateway

### Issues

Fixes #1561

### Automation tests
 - Unit tests added: Yes
 - Integration tests added: No

### Tested environments

Ubuntu 20.04.1

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [ ] Validated respective github issues
- [ ] Assigned milestone to the github issue(s)

### CODE REVIEW

File: adapter/pkg/oasparser/envoyconf/routesWithClusters.go [hunk 1]
```diff
@@ -86,6 +87,13 @@ func CreateRoutesWithClusters(mgwSwagger model.MgwSwagger, upstreamCerts []byte)
 		endpoints = append(endpoints, apilevelAddressP)
 		apiEndpointBasePath = apiLevelEndpointProd[0].Basepath
 
+		//todo add to ClusterConsulKeyMap on Sand , Resource level as well
```
Comment (menakaj) [inline adapter/pkg/oasparser/envoyconf/routesWithClusters.go:90]: Is this todo still valid?

File: adapter/pkg/svcdiscovery/client.go [hunk 2]
```diff
@@ -0,0 +1,250 @@
+/*
+ *  Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (menakaj) [inline adapter/pkg/svcdiscovery/client.go:2]: Make the license year as 2021

Comment (menakaj) [reply adapter/pkg/svcdiscovery/client.go:2]: Fix in other files as well

File: adapter/pkg/svcdiscovery/utils.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+//IsDiscoveryServiceEndpoint checks whether an endpoint string is a consul syntax string
+func IsDiscoveryServiceEndpoint(str string) bool {
+	str = strings.TrimSpace(str)
+	re, _ := regexp.Compile(`^consul(\s*)\(.*,.*\)$`)
+	return re.MatchString(str)
+}
+
+//parse a list of datacenters or tags
+func parseList(str string) []string {
+	s := strings.Split(str, ",")
```
Comment (menakaj) [inline adapter/pkg/svcdiscovery/utils.go:47]: Shall we use proper variable names?

File: adapter/pkg/xds/xds_server.go [hunk 4]
```diff
@@ -274,3 +279,132 @@ func updateXdsCache(label string, endpoints []types.Resource, clusters []types.R
...
+	mutexForXdsUpdate.Lock()
+	defer mutexForXdsUpdate.Unlock()
+	updateXdsCache(label, endpoints, clusters, routes, listeners)
+}
+
+func startConsulServiceDiscovery() {
+	//label := "default"
+	for apiKey, clusterList := range openAPIClustersMap {
+		//logger.LoggerXds.Debugln("API key for consul service discovery: ", apiKey)
+		//logger.LoggerXds.Info(svcdiscovery.ClusterConsulKeyMap)
```
Comment (menakaj) [inline adapter/pkg/xds/xds_server.go:295]: Remove commented code

File: adapter/pkg/xds/xds_server.go [hunk 5]
```diff
@@ -274,3 +279,132 @@ func updateXdsCache(label string, endpoints []types.Resource, clusters []types.R
...
+					logger.LoggerXds.Error("consul syntax parse error ", errConSyn)
+					return
+				}
+				logger.LoggerXds.Info(query)
+				go getServiceDiscoveryData(query, cluster.Name, apiKey)
+			}
+		}
+
+	}
+
```
Comment (menakaj) [inline adapter/pkg/xds/xds_server.go:311]: Remove extra line

File: adapter/pkg/xds/xds_server.go [hunk 6]
```diff
@@ -274,3 +279,132 @@ func updateXdsCache(label string, endpoints []types.Resource, clusters []types.R
...
+				query, errConSyn := svcdiscovery.ParseQueryString(consulSyntax)
+				if errConSyn != nil {
+					logger.LoggerXds.Error("consul syntax parse error ", errConSyn)
+					return
+				}
+				logger.LoggerXds.Info(query)
+				go getServiceDiscoveryData(query, cluster.Name, apiKey)
+			}
+		}
+
```
Comment (menakaj) [inline adapter/pkg/xds/xds_server.go:309]: Remove extra line

File: adapter/pkg/xds/xds_server.go [hunk 7]
```diff
@@ -274,3 +279,132 @@ func updateXdsCache(label string, endpoints []types.Resource, clusters []types.R
...
+	for {
+		select {
+		case queryResultsList, ok := <-resultChan:
+			if !ok { //ok==false --> result chan is closed
+				logger.LoggerXds.Info("closed the result channel for cluster name: ", clusterName)
+				return
+			}
+			//logger.LoggerXds.Info(queryResultsList)
+
+			//logger.LoggerXds.Info("map: ", svcdiscovery.ClusterConsulResultMap)
```
Comment (menakaj) [inline adapter/pkg/xds/xds_server.go:328]: Remove commented code

Comment (menakaj) [reply adapter/pkg/xds/xds_server.go:328]: Fix in other places in this file as well.

File: adapter/pkg/svcdiscovery/client.go (hunk 2 above)
Comment (rumeshmadhusanka) [reply adapter/pkg/svcdiscovery/client.go]: fixed 8351704

File: adapter/pkg/svcdiscovery/utils.go (hunk 3 above)
Comment (rumeshmadhusanka) [reply adapter/pkg/svcdiscovery/utils.go]: fixed 1a202af

File: adapter/pkg/oasparser/envoyconf/routesWithClusters.go (hunk 1 above)
Comment (rumeshmadhusanka) [reply adapter/pkg/oasparser/envoyconf/routesWithClusters.go]: removed outdated todo 
6664e33

File: adapter/pkg/xds/xds_server.go (hunk 7 above)
Comment (rumeshmadhusanka) [reply adapter/pkg/xds/xds_server.go]: fixed 5d85993

File: adapter/pkg/oasparser/model/mgwSwagger.go [hunk 8]
```diff
@@ -156,12 +159,32 @@ func getXWso2Endpoints(vendorExtensible map[string]interface{}, endpointType str
...
 				for _, v := range castedUrlsInterface {
-					endpoint := getHostandBasepathandPort(v.(string))
-					endpointType, endpointTypeFound := val[typeConst]
-					if endpointTypeFound {
-						endpoint.URLType = endpointType.(string)
+					if svcdiscovery.IsDiscoveryServiceEndpoint(v.(string)) {
+						logger.LoggerOasparser.Debug("consul query syntax found: ", v.(string))
+						queryString, defHost, err := svcdiscovery.ParseConsulSyntax(v.(string))
+						if err != nil {
+							logger.LoggerOasparser.Error("consul syntax parse error")
```
Comment (Rajith90) [inline adapter/pkg/oasparser/model/mgwSwagger.go:166]: Lets print the error as well

File: adapter/pkg/svcdiscovery/client.go [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	logger "github.com/wso2/micro-gw/loggers"
+	"io/ioutil"
+	"net/http"
+	"time"
+	//logger "github.com/wso2/micro-gw/loggers"
```
Comment (Rajith90) [inline adapter/pkg/svcdiscovery/client.go:28]: Remove commented lines

File: adapter/pkg/svcdiscovery/client.go [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+		url += "/"
+	}
+	url += apiPath
+	url += path
+
+	req, _ := http.NewRequest("GET", url, nil)
+
+	//add query parameters
+	q := req.URL.Query()
+	q.Add("dc", dc)       //datacenter
```
Comment (Rajith90) [inline adapter/pkg/svcdiscovery/client.go:140]: Lets use constants

File: adapter/pkg/svcdiscovery/client.go [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+	q.Add("passing", "1") // health checks passing only
```
Comment (Rajith90) [inline adapter/pkg/svcdiscovery/client.go:141]: Lets use constants

File: adapter/pkg/svcdiscovery/client.go [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+	if namespace != "" {  //namespace, an enterprise feature
+		q.Add("nc", namespace)
```
Comment (Rajith90) [inline adapter/pkg/svcdiscovery/client.go:143]: Lets use constants

File: adapter/pkg/svcdiscovery/client.go [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+	}
+
+	req.URL.RawQuery = q.Encode()
+	response, errHTTP := c.client.Do(req)
+	if errHTTP != nil {
+		return []Upstream{}, errHTTP
+	}
+	//set headers
+	req.Header.Set("X-Consul-Token", c.aclToken)
```
Comment (Rajith90) [inline adapter/pkg/svcdiscovery/client.go:152]: Lets use constants for X-Consul-Token

File: adapter/pkg/svcdiscovery/client.go [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+}
+
+//gets the upstreams for single query(ex: [dc1,dc2].namespace.serviceA.[tag1,tag2])
+//there can be many upstreams per query
+//sends the respective upstreams through resultChan
+func (c ConsulClient) getUpstreams(query Query, resultChan chan []Upstream) {
+
+	defer func() {
+		if r := recover(); r != nil {
+			logger.LoggerSvcDiscovery.Error("Recovered of from a panic: ", r)
```
Comment (Rajith90) [inline adapter/pkg/svcdiscovery/client.go:186]:
```suggestion
			logger.LoggerSvcDiscovery.Error("Recovered from a panic: ", r)
```

File: adapter/pkg/svcdiscovery/client.go [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+		}
+	}()
+
+	var result []Upstream
+	for _, dc := range query.Datacenters {
+		res, errGet := c.get(query.ServiceName, dc, query.Namespace, query.Tags)
+		if errGet == nil {
+			result = append(result, res...)
+		} else {
+			logger.LoggerSvcDiscovery.Error("service registry unreachable ", errGet)
```
Comment (Rajith90) [inline adapter/pkg/svcdiscovery/client.go:197]:
```suggestion
			logger.LoggerSvcDiscovery.Error("Service registry unreachable ", errGet)
```

File: adapter/pkg/svcdiscovery/client.go [hunk 16]
```diff
@@ -0,0 +1,249 @@
...
+		}
+	}
+
+	if len(result) == 0 {
+		logger.LoggerSvcDiscovery.Error("consul service registry query came up with empty result/ service registry unreachable")
```
Comment (Rajith90) [inline adapter/pkg/svcdiscovery/client.go:202]:
```suggestion
		logger.LoggerSvcDiscovery.Error("Consul service registry query came up with empty result/ service registry unreachable")
```

Comment (Rajith90) [reply adapter/pkg/svcdiscovery/client.go:202]: result length zero does not mean that "Service registry unreachable". Does it?

File: adapter/pkg/svcdiscovery/client.go [hunk 17]
```diff
@@ -0,0 +1,249 @@
...
+
+//Poll periodically poll consul for updates using getUpstreams() func.
+//doneChan is there to release resources
+//closing the doneChan will stop polling
+func (c ConsulClient) Poll(query Query, doneChan <-chan bool) <-chan []Upstream {
+	resultChan := make(chan []Upstream)
+
+	//do not start polling consul if there are config errors
+	if errConfLoad != nil {
+		logger.LoggerSvcDiscovery.Error("Config errors found in consul config. ", query, " wil not be polled")
```
Comment (Rajith90) [inline adapter/pkg/svcdiscovery/client.go:217]: Lets print the error as well

File: adapter/pkg/svcdiscovery/client.go [hunk 18]
```diff
@@ -0,0 +1,249 @@
...
+		}()
+		// release resources when this go routine exits
+		defer close(resultChan)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-doneChan:
+				//sending a signal through doneChan will cause this go routine to exit
+				logger.LoggerSvcDiscovery.Info("consul query stopped polling for:", query)
```
Comment (Rajith90) [inline adapter/pkg/svcdiscovery/client.go:240]:
```suggestion
				logger.LoggerSvcDiscovery.Info("Consul query stopped polling for:", query)
```

File: resources/conf/config.toml [hunk 19]
```diff
@@ -9,6 +9,15 @@ privateKeyPath = "/home/wso2/security/localhost.key"
...
 password = "admin"
 
+[adapter.consul]
+url = "https://169.254.1.1:8501" # scheme + host ip + port
+pollInterval = 5 # seconds
+aclTokenFilePath = ""
+# certs for tls
+caCertPath = "/home/wso2/security/consul/consul-agent-ca.pem"
+certPath = "/home/wso2/security/consul/local-dc-client-consul-0.pem"
+keyPath = "/home/wso2/security/consul/local-dc-client-consul-0-key.pem"
```
Comment (Rajith90) [inline resources/conf/config.toml:19]: Make a note to check the certificates

Comment (rumeshmadhusanka) [reply resources/conf/config.toml:20]: fixed 9717e0a

File: adapter/pkg/svcdiscovery/client.go (hunk 17 above)
Comment (rumeshmadhusanka) [reply adapter/pkg/svcdiscovery/client.go]: fixed 0ad0d8c

File: adapter/pkg/svcdiscovery/client.go (hunk 13 above)
Comment (rumeshmadhusanka) [reply adapter/pkg/svcdiscovery/client.go]: fixed 0ad0d8c

File: adapter/pkg/svcdiscovery/client.go (hunk 11 above)
Comment (rumeshmadhusanka) [reply adapter/pkg/svcdiscovery/client.go]: fixed 0ad0d8c

File: adapter/pkg/svcdiscovery/client.go (hunk 10 above)
Comment (rumeshmadhusanka) [reply adapter/pkg/svcdiscovery/client.go]: fixed 0ad0d8c

File: adapter/pkg/svcdiscovery/client.go (hunk 9 above)
Comment (rumeshmadhusanka) [reply adapter/pkg/svcdiscovery/client.go]: fixed 0ad0d8c

File: adapter/pkg/oasparser/model/mgwSwagger.go (hunk 8 above)
Comment (rumeshmadhusanka) [reply adapter/pkg/oasparser/model/mgwSwagger.go]: fixed 0ad0d8c

**Review by Rajith90 — APPROVED** [review]

**Pull request merged**